homeworks/09/chat-system/src/Server.java

```diff
@@ -0,0 +1,149 @@
+        for (ConnectionHandler ch : connections) {
+            if (ch != null) {
+                if (ch.threadSender == false) {
+                    ch.sendMessage(message);
+                }
+            }
+        }
+    }
+
+    public void shutdown() {
+        try {
+            done = true;
+            if (!server.isClosed()) {
+                server.close();
+            }
+            for (ConnectionHandler ch : connections) {
+
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    class ConnectionHandler implements Runnable {
+        private Socket client;
+        private BufferedReader input;
+        private PrintWriter output;
+        private String nickname;
+        private boolean threadSender;
+
+        public ConnectionHandler(Socket client) {
+            this.client = client;
+        }
+
+        @Override
+        public void run() {
+            try {
+                output = new PrintWriter(client.getOutputStream(), true);
+                input = new BufferedReader(new InputStreamReader(client.getInputStream()));
+                nickname = "User";
+                System.out.println("New " + nickname + " connected!");
+                broadcast("New " + nickname + " joined the chat!");
+                String message;
+                while ((message = input.readLine()) != null) {
+                    if (message.startsWith("/nick")) {
+                        String[] splittedMessage = message.split(" ", 2);
+                        if (splittedMessage.length == 2) {
+                            threadSender = true;
+                            broadcast(nickname + " renamed themselves to " + splittedMessage[1]);
+                            threadSender = false;
+                            System.out.println(nickname + " renamed themselves to " + splittedMessage[1]);
+                            nickname = splittedMessage[1];
+                            output.println("Successfully changed nickname to " + nickname);
+                        } else {
+                            output.println("No nickname provided!");
+                        }
+                    } else if (message.startsWith("/msg")){
+                        String[] splittedMessage = message.split(" ", 2);
+                        LocalDateTime now = LocalDateTime.now();
+                        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm a");
+                        threadSender = true;
+                        broadcast("[" + now.format(formatter) + "] " + nickname + ": " + splittedMessage[1]);
+                        threadSender = false;
+                    } else if (message.startsWith("/quit")) {
+                        threadSender = true;
+                        System.out.println(nickname + " left the chat!");
+                        broadcast(nickname + " left the chat!");
+                        threadSender = false;
```

Comment:
Using a boolean variable (threadSender) to check who is a message's original sender may lead to issues if multiple people send messages at once. I recommend to just pass a reference of the **ClientHandler** in the **Server**'s **broadcast** method to check if the client should receive a message.